### PR TITLE
Add custom destroyerIdentity parameter to the sendBlockBreak function

### LINKED ISSUE #5246: No multiple different block-damage to a player possible
### What behaviour is expected:
Multiple blocks having break-texture/design/overlay

### What behaviour is observed:
only one block a time has the break-texture/design/overlay

### Steps/models to reproduce:
Creating multiple different timed (async) BukkitRunnable's sendBlockDamage different Blocks (at different times) to the same player

### Plugin list:
Our Framework, our Plugin, LuckPerms, Multiverse-Core, ProtocolSupport, WorldEdit

### Paper version:
git-Paper-496 1.16.5-R0.1

### Anything else:
```kotlin
it.sendBlockDamage(block.location, (attempt.toDouble() / (remainingSeconds*timeMultiplikator)).toFloat())
```

The BlockDamage of an another block is dissappearing when an other block is receiving an block-damage

### PR DESCRIPTION
**Pull request by TheFruxz**
This solves the issue #5246

### CODE REVIEW

File: Spigot-Server-Patches/0760-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index cfe6898dc373fe55a08acf5c90e200061aa7d0fc..98d5945140912e75c5a915823ea5e9c37cf6db10 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -137,6 +137,7 @@ import org.bukkit.plugin.messaging.StandardMessenger;
+ import org.bukkit.scoreboard.Scoreboard;
+ 
+ import net.md_5.bungee.api.chat.BaseComponent; // Spigot
++import org.jetbrains.annotations.NotNull;
```
Comment (Machine-Maker) [inline Spigot-Server-Patches/0760-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch]: Don't import here, nullability annotations aren't needed for the craftbukkit impl of the API, so you can either just remove it, or use the fully qualified name below.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Comment by TheFruxz** [comment]
@Machine-Maker Changes done!

**Review by Machine-Maker — APPROVED** [review]

File: Spigot-API-Patches/0318-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+     public void sendBlockDamage(@NotNull Location loc, float progress);
+ 
++    // Paper start
++    /**
++     * Send block damage. This fakes block break progress for a user at a
++     * certain location. This will not actually change the block's break
++     * progress in any way.
++     *
++     * @param loc the location of the damaged block
++     * @param destroyerIdentity the entity-id of the destroyer, can be custom to display multiple block-damages
```
Comment (kennytv) [inline Spigot-API-Patches/0318-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch]:
```suggestion
+     * @param destroyerIdentity the entity id of the destroyer
```

File: Spigot-Server-Patches/0760-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
++        Preconditions.checkArgument(loc != null, "loc must not be null");
++        Preconditions.checkArgument(progress >= 0.0 && progress <= 1.0, "progress must be between 0.0 and 1.0 (inclusive)");
++
++        if (getHandle().playerConnection == null) return;
++
++        int stage = (int) (9 * progress); // There are 0 - 9 damage states
++        PacketPlayOutBlockBreakAnimation packet = new PacketPlayOutBlockBreakAnimation(destroyerIdentity, new BlockPosition(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);
++        getHandle().playerConnection.sendPacket(packet);
++    }
++
```
Comment (kennytv) [inline Spigot-Server-Patches/0760-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch]: Missing `// Paper end`

Comment (kennytv) [inline Spigot-Server-Patches/0760-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch]: Don't copy the entire method for a one-line change inside of it; add the id to the existing method, and add a new method with the old parameters that calls `sendBlockDamage(loc, this.getHandle().getId(), progress)`

**Review by kennytv — DISMISSED** [review]
Also has to be rebased/updated to Mojang mappings

File: Spigot-Server-Patches/0760-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch (hunk 3 above)
Comment (NoahvdAa) [reply Spigot-Server-Patches/0760-Add-custom-destroyerIdentity-int-parameter-to-the-se.patch]: The `// Paper start` comment seems to be from a different patch (Adventure?), which probably also has the `// Paper end` comment.

**Comment by TheFruxz** [comment]
everything should be in order now, i hope :D

- [x] applied changes
- [x] "upgraded" to new structure

File: patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ 
+         if (this.getHandle().connection == null) return;
+ 
+         int stage = (int) (9 * progress); // There are 0 - 9 damage states
+-        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(this.getHandle().getId(), new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);
++        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(destroyerIdentity, new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);
+         this.getHandle().connection.send(packet);
+     }
+ 
+-    // Paper start
```
Comment (Machine-Maker) [inline patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch]: Don't remove this comment

File: patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+index c64911651f3d736c83cc83996de04920b091cc57..65d600f570db8b1308ce0a44ebdf970cface4518 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -845,19 +845,24 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+         this.getHandle().connection.send(packet);
+     }
+ 
++    // Paper start
```
Comment (Machine-Maker) [inline patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch]: Don't start the // Paper comment block here, start it right above the first line changed

File: patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
++    @Override
++    public void sendBlockDamage(Location loc, int destroyerIdentity, float progress) {
+         Preconditions.checkArgument(loc != null, "loc must not be null");
+         Preconditions.checkArgument(progress >= 0.0 && progress <= 1.0, "progress must be between 0.0 and 1.0 (inclusive)");
+ 
+         if (this.getHandle().connection == null) return;
+ 
+         int stage = (int) (9 * progress); // There are 0 - 9 damage states
+-        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(this.getHandle().getId(), new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);
++        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(destroyerIdentity, new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);
```
Comment (Machine-Maker) [inline patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch]: Add inline // Paper comment

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

**Comment by TheFruxz** [comment]
Applied changes

**Comment by TheFruxz** [comment]
I think that a rebased is now again required

File: patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+@@ -847,15 +847,22 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+ 
+     @Override
+     public void sendBlockDamage(Location loc, float progress) {
++        sendBlockDamage(loc, this.getHandle().getId(), progress); // Paper - customBlockDamage identity
++    }
++
++    // Paper start - customBlockDamage identity
++    @Override
++    public void sendBlockDamage(Location loc, int destroyerIdentity, float progress) {
```
Comment (Owen1212055) [inline patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch]:
```suggestion
+        sendBlockDamage(loc, this.getHandle().getId(), progress); // Paper start - customBlockDamage identity
+    }
+
+    @Override
+    public void sendBlockDamage(Location loc, int destroyerIdentity, float progress) { // Paper end - customBlockDamage identity
```

File: patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+         Preconditions.checkArgument(loc != null, "loc must not be null");
+         Preconditions.checkArgument(progress >= 0.0 && progress <= 1.0, "progress must be between 0.0 and 1.0 (inclusive)");
+ 
+         if (this.getHandle().connection == null) return;
+ 
+         int stage = (int) (9 * progress); // There are 0 - 9 damage states
+-        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(this.getHandle().getId(), new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);
++        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(destroyerIdentity, new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage);
```
Comment (Owen1212055) [inline patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch]:
```suggestion
+        ClientboundBlockDestructionPacket packet = new ClientboundBlockDestructionPacket(destroyerIdentity, new BlockPos(loc.getBlockX(), loc.getBlockY(), loc.getBlockZ()), stage); // Paper - customBlockDamage identity
```

File: patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+         this.getHandle().connection.send(packet);
+     }
++    // Paper end - customBlockDamage identity
```
Comment (Owen1212055) [inline patches/server/0911-Add-custom-destroyerIdentity-parameter-to-the-sendBl.patch]: Remove this paper comment

**Review by Owen1212055 — CHANGES_REQUESTED** [review]
Your paper comments don't represent where you are actually modifying the file. You are supposed to wrap your dif in the paper comments, not really just mark the entire method.

I gave you some pointers here, be sure to modify it yourself (don't commit it directly!)

**Comment by u9g** [comment]
Do you intend to finish this pr?

**Comment by TheFruxz** [comment]
> Do you intend to finish this pr?

Thanks for the reminder, have already forgotten about this PR.

Yes, I will take care of the requested changes soon.

**Comment by TheFruxz** [comment]
Changes applied

File: patches/server/0981-Add-custom-destroyerIdentity-to-sendBlockDamage.patch [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+@@ -917,13 +917,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
+ 
+     @Override
+     public void sendBlockDamage(Location loc, float progress) {
++        sendBlockDamage(loc, progress, this.getHandle().getId()); // Paper start - customBlockDamage identity
++    }
++
++    @Override
++    public void sendBlockDamage(Location loc, float progress, int destroyerIdentity) { // Paper end - customBlockDamage identity
```
Comment (Machine-Maker) [inline patches/server/0981-Add-custom-destroyerIdentity-to-sendBlockDamage.patch]: Throw those paper start and end comments above and below those lines respectively. Basically, you want the `// Paper start` and `// Paper end` comments to perfectly wrap the diff when you look at this patch, so the first line of the diff should have the start comment only, and the last has the end comment only.

File: patches/api/0393-Add-custom-destroyerIdentity-to-sendBlockDamage.patch [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
++
++    /**
++     * Send block damage. This fakes block break progress for a user at a certain location.
++     * This will not actually change the block's break progress in any way.
++     *
++     * @param loc the location of the damaged block
++     * @param progress the progress from 0.0 - 1.0 where 0 is no damage and 1.0 is the most damaged
++     * @param destroyerIdentity the entity id of the destroyer
++     */
++    public void sendBlockDamage(@NotNull Location loc, float progress, int destroyerIdentity);
```
Comment (Machine-Maker) [inline patches/api/0393-Add-custom-destroyerIdentity-to-sendBlockDamage.patch]: Perhaps some additional javadoc comment should be added to the original sendBlockDamage telling them that the destroying will be this player.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

File: patches/api/0393-Add-custom-destroyerIdentity-to-sendBlockDamage.patch (hunk 11 above)
Comment (TheFruxz) [reply patches/api/0393-Add-custom-destroyerIdentity-to-sendBlockDamage.patch]: If I want to change the javadoc of this method of the interface, should I move the whole method to the `// Paper start` part?

File: patches/server/0981-Add-custom-destroyerIdentity-to-sendBlockDamage.patch (hunk 10 above)
Comment (TheFruxz) [reply patches/server/0981-Add-custom-destroyerIdentity-to-sendBlockDamage.patch]: now seperate lines

File: patches/api/0393-Add-custom-destroyerIdentity-to-sendBlockDamage.patch (hunk 11 above)
Comment (Machine-Maker) [reply patches/api/0393-Add-custom-destroyerIdentity-to-sendBlockDamage.patch]: no, just don't add a // Paper comment. javadoc changes don't require a comment for them.

Comment (TheFruxz) [reply patches/api/0393-Add-custom-destroyerIdentity-to-sendBlockDamage.patch]: Added javadoc comments

**Review by Machine-Maker — APPROVED** [review]

**Review by Owen1212055 — APPROVED** [review]

**Review by electronicboy — APPROVED** [review]

**Review by lynxplay — APPROVED** [review]

**Pull request merged**